add VPC docs and improve ELB docs (#1584)
SigNoz/signoz.io: docs/aws-monitoring/vpc-logs.md

</figure>
<br />

**Step 4:** Select `Create a new role with basic Lambda permissions`for now, we’ll requiring more permissions down the lane. So for now, select this option. 

<figure data-zoomable align='center'>
    <img className="box-shadowed-image" src="/img/docs/elb/elb-logs-lambda_4.webp" alt=""/>
    <figcaption><i>Your barebones Lambda function is created now</i></figcaption>
</figure>
<br />

Once you are done configuring the lambda function, you’ll see it get created here. Feel free to move around and see what all things AWS Lambda offers. 

### Configuring Policies for Lambda function

As said in Step 4 previously, we need extra permissions in order to access the S3 Bucket for execution of our Lambda code, follow along to set it up. 

**Step 1:** Scroll down from your Lambda page, you’ll see a few tabs there. Go to `Configurations` and select `Permissions` from the left sidebar. 

<figure data-zoomable align='center'>
    <img className="box-shadowed-image" src="/img/docs/elb/elb-logs-lambda_5.webp" alt=""/>
    <figcaption><i>Choosing execution role from Configurations tab</i></figcaption>
</figure>
<br />

> **Step 2:** Click on the `Execution Role name`'s link, it will take us to AWS IAM page. Here, we’ll be a policy to get full S3 access.

Author: As I have already annotated the image with an arrow, giving the link name again will be redundant. If there were no arrows in the image, your comment was meaningful to implement.

To setup the trigger, follow these steps:

**Step 1:** Click on the `+ Add trigger`button from the Lambda console. 

<figure data-zoomable align='center'>
    <img className="box-shadowed-image" src="/img/docs/elb/elb-logs-lambda_8.webp" alt=""/>
    <figcaption><i>Add a trigger to the function.</i></figcaption>
</figure>
<br />

**Step 2:** Select `S3` from the first drop down of AWS services list. Pick you S3 bucket for the second field. 

**Step 3:** For the Event types field, you can select any number of options you wish. The trigger will occur depending upon what option(s) you choose here. By default, the `All object create events` will be selected. 

<figure data-zoomable align='center'>
    <img className="box-shadowed-image" src="/img/docs/elb/elb-logs-lambda_9.webp" alt=""/>
    <figcaption><i>Choose event types from the drop down menu</i></figcaption>
</figure>
<br />

Verify the settings and click on `Add` button at bottom right to add this trigger. You could add more than one trigger, but we’ll stick to just one for now. 

### Adding Request Layer

> We will be using python’s `request` module and by default, you do not get ‘requests’ module by default in the Lambda. Why? See the attached image and [this link](https://aws.amazon.com/blogs/compute/upcoming-changes-to-the-python-sdk-in-aws-lambda/). 

Reviewer: ```suggestion
We will be using python's `request` module which is [not included by default](https://aws.amazon.com/blogs/compute/upcoming-changes-to-the-python-sdk-in-aws-lambda/) in Lambda.
```

> <figure data-zoomable align='center'>

Reviewer: We can remove this image - users can check out the link above if they want more details.

    <img className="box-shadowed-image" src="/img/docs/elb/elb-logs-upcoming-changes.webp" alt=""/>
    <figcaption><i>AWS Lambda changes for Python SDK</i></figcaption>
</figure>
<br />

To utilize Python's `requests` module within a Lambda function, it's necessary to explicitly add it as a layer. While there may be alternative approaches, it's advisable to adhere to established practices that have been thoroughly tested and proven effective. Therefore, we will proceed with adding `requests` as a layer to ensure reliable functionality within the Lambda environment.

> Anyways, refer to the below attached steps to create a zip of the request module and add it as a layer to make it work on AWS lambda. Steps for which are described ahead. 

Reviewer: ```suggestion
Follow the steps below to create a zip of the request module and add it as a layer to make it work on AWS lambda. 
```

> The commands you’d need:

Reviewer: Can we make this step 1 ?

```py
# make a new directory
mkdir python
# move into that directory
cd python

# install requests module
pip install --target . requests
# zip the contents under the name dependencies.zip
zip -r dependencies.zip ../python 

```

**Step 1:** To upload your zip file, go to AWS Lambda > Layers and click on `Create Layer`. [Not inside your specific Lambda function, just the landing page of AWS Lambda]. 

<figure data-zoomable align='center'>
    <img className="box-shadowed-image" src="/img/docs/elb/elb-logs-layer_3.webp" alt=""/>
    <figcaption><i>Creating a new Layer</i></figcaption>
</figure>
<br />

**Step 2:** You’ll be redirected to Layer configurations page, here, give a name to your layer, an optional description, select `Upload a .zip file` , click on `Upload` and locate the requirements.zip file. 

> **Step 3:** Select your desired architecture and pick `Python 3.x` as your runtime. Hit `Create`. You layer as now been created. Now lets connect it to our Lambda function which we created to send logs to SigNoz. 

Reviewer: ```suggestion
**Step 3:** Select your desired architecture and pick `Python 3.x` as your runtime. Hit `Create`. Your layer has now been created. Now lets connect it to our Lambda function which we created to send logs to SigNoz. 
```

1. Assuming all standard log lines have a timestamp field.
2. Parse and select the timestamp field from the log line and add it before the `response = requests.post(http_url, json=json_data)` line as a if else condition to only send logs which are x days older (say 3 days). 

So, the function now will first check the log timestamp and only send those logs which are 3 days older (say) or even a few hours old. 

Let’s consider the below pseudo code for better understanding:

```python
   from datetime import datetime, timedelta

   # Your given timestamp
   given_timestamp_str = "2024-01-01T23:58:02.231919Z"
   given_timestamp = datetime.fromisoformat(given_timestamp_str.replace('Z', '+00:00'))

   # Current time
   current_time = datetime.utcnow()

   # Calculate the time difference
   time_difference = current_time - given_timestamp

   # Check if the time difference is less than 3 days
   if time_difference < timedelta(days=3):
       # Run your specific function here
       print("Running the specific function.")
   # ADD THE response = requests.post(http_url, json=json_data) LINE HERE

   else:
       print("Time difference exceeds 3 days. Function will not run.")
```

> Feel free to modify any part of the code if you wish to. 

Reviewer: ```suggestion
Feel free to modify any part of the code according to your requirements.
```
